lib/backend/k8s/resources/profile.go

```diff
@@ -148,7 +251,212 @@ func (c *profileClient) Watch(ctx context.Context, list model.ListInterface, rev
 		if !ok {
 			return nil, errors.New("profile conversion with incorrect k8s resource type")
 		}
-		return c.converter.NamespaceToProfile(k8sNamespace)
+		return c.NamespaceToProfile(k8sNamespace)
 	}
-	return newK8sWatcherConverter(ctx, "Profile", converter, k8sWatch), nil
+
+	nsWatcher := newK8sWatcherConverter(ctx, "Profile-NS", converter, nsWatch)
+
+	if c.AlphaSA == false {
+		return nsWatcher, nil
+	}
+
+	// Watch all service accounts in ALL namespaces
+	saWatch, err := c.clientSet.CoreV1().ServiceAccounts(kapiv1.NamespaceAll).Watch(metav1.ListOptions{ResourceVersion: saRev})
+	if err != nil {
+		nsWatch.Stop()
+		return nil, err
+	}
+
+	converterSA := func(r Resource) (*model.KVPair, error) {
+		k8sServiceAccount, ok := r.(*kapiv1.ServiceAccount)
+		if !ok {
+			nsWatch.Stop()
+			return nil, errors.New("Profile converion with incorrect k8s resource type")
+		}
+		return c.ServiceAccountToProfile(k8sServiceAccount)
+	}
+
+	saWatcher := newK8sWatcherConverter(ctx, "Profile-SA", converterSA, saWatch)
+
+	return newProfileWatcher(ctx, nsWatcher, saWatcher), nil
+
+}
+
+func newProfileWatcher(ctx context.Context, k8sWatchNS, k8sWatchSA api.WatchInterface) api.WatchInterface {
+	ctx, cancel := context.WithCancel(ctx)
+	wc := &profileWatcher{
+		k8sNSWatch: k8sWatchNS,
+		k8sSAWatch: k8sWatchSA,
+		context:    ctx,
+		cancel:     cancel,
+		resultChan: make(chan api.WatchEvent, resultsBufSize),
+	}
+	go wc.processProfileEvents()
+	return wc
+}
+
+type profileWatcher struct {
+	conversion.Converter
+	converter  ConvertK8sResourceToKVPair
+	k8sNSWatch api.WatchInterface
+	k8sSAWatch api.WatchInterface
+	k8sNSRev   string
+	k8sSARev   string
+	context    context.Context
+	cancel     context.CancelFunc
+	resultChan chan api.WatchEvent
+	terminated uint32
+}
+
+// Stop stops the watcher and releases associated resources.
+// This calls through the context cancel function.
+func (pw *profileWatcher) Stop() {
+	pw.cancel()
+	pw.k8sNSWatch.Stop()
+	pw.k8sSAWatch.Stop()
+}
+
+// ResultChan returns a channel used to receive WatchEvents.
+func (pw *profileWatcher) ResultChan() <-chan api.WatchEvent {
+	return pw.resultChan
+}
+
+// HasTerminated returns true when the watcher has completed termination processing.
+func (pw *profileWatcher) HasTerminated() bool {
+	terminated := atomic.LoadUint32(&pw.terminated) != 0
+
+	if pw.k8sNSWatch != nil {
+		terminated = terminated && pw.k8sNSWatch.HasTerminated()
+	}
+	if pw.k8sSAWatch != nil {
+		terminated = terminated && pw.k8sSAWatch.HasTerminated()
+	}
+
+	return terminated
+}
+
+// Loop to process the events stream from the underlying k8s Watcher and convert them to
+// backend KVPs.
+func (pw *profileWatcher) processProfileEvents() {
```

Comment:
Seems fine for alpha, but I suspect we're duplicating a lot of code between this and the NP stuff?

Might want to revisit commonizing some of this when we move out of alpha. @robbrockbank wdyt?

Reply:
most of this will vanish once we start using API groups to handle this - that's why i thought this was fine to leave as is for now.